File: python/example_tts_long_input.py
#!/usr/bin/env python3
"""
Example script for Inworld TTS synthesis with long text input.

This script demonstrates how to synthesize speech from long text by:
1. Chunking text at sentence boundaries (after ~1000 characters)
2. Processing each chunk through the TTS API in parallel
3. Stitching all audio outputs together
4. Reporting splice points with timestamps for quality checking
"""

import base64
import os
import re
import time
import wave
from concurrent.futures import ThreadPoolExecutor, as_completed
from dataclasses import dataclass
from pathlib import Path

import requests

# Configuration
MIN_CHUNK_SIZE = 1000  # Minimum characters before looking for sentence end
MAX_CHUNK_SIZE = 1900  # Maximum chunk size (API limit is 2000)
SAMPLE_RATE = 48000
BITS_PER_SAMPLE = 16
CHANNELS = 1


@dataclass
class TextChunk:
    """Represents a chunk of text with its position in the original text."""
    text: str
    start_char: int
    end_char: int


@dataclass
class SplicePoint:
    """Represents a splice point in the combined audio."""
    splice_index: int
    timestamp: float
    formatted_time: str
    chunk_start_char: int
    chunk_end_char: int
    text_preview: str


def check_api_key():
    """Check if INWORLD_API_KEY environment variable is set."""
    api_key = os.getenv("INWORLD_API_KEY")
    if not api_key:
        print("Error: INWORLD_API_KEY environment variable is not set.")
        print("Please set it with: export INWORLD_API_KEY=your_api_key_here")
        return None
    return api_key


def chunk_text(text: str) -> list[TextChunk]:
    """
    Chunk text into segments at sentence boundaries.
    After MIN_CHUNK_SIZE characters, looks for sentence-ending punctuation.
    
    Args:
        text: The full text to chunk
        
    Returns:
        List of TextChunk objects with text and positions
    """
    chunks = []
    current_position = 0
    
    # Regex for sentence-ending punctuation
    sentence_end_pattern = re.compile(r"[.!?][\"'\"']?\s+|[.!?][\"'\"']?$")
    
    while current_position < len(text):
        remaining_text = text[current_position:]
        
        # If remaining text is short enough, take it all
        if len(remaining_text) <= MAX_CHUNK_SIZE:
            chunk_text = remaining_text.strip()
            if chunk_text:
                chunks.append(TextChunk(
                    text=chunk_text,
                    start_char=current_position,
                    end_char=len(text)
                ))
            break
        
        # Look for sentence end after MIN_CHUNK_SIZE
        search_text = remaining_text[:MAX_CHUNK_SIZE]
        chunk_end = -1
        
        # Find all sentence endings in the search range after MIN_CHUNK_SIZE
        matches = list(sentence_end_pattern.finditer(search_text))
        valid_matches = [m for m in matches if m.end() >= MIN_CHUNK_SIZE]
        
        if valid_matches:
            # Use the first sentence end after MIN_CHUNK_SIZE
            chunk_end = valid_matches[0].end()
        elif matches:
            # Use the last sentence end before MAX_CHUNK_SIZE
            chunk_end = matches[-1].end()
        else:
            # No sentence end found, break at last space
            space_index = search_text.rfind(' ')
            chunk_end = space_index + 1 if space_index > 0 else MAX_CHUNK_SIZE
        
        chunk_text = remaining_text[:chunk_end].strip()
        if chunk_text:
            chunks.append(TextChunk(
                text=chunk_text,
                start_char=current_position,
                end_char=current_position + chunk_end
            ))
        
        current_position += chunk_end
    
    return chunks


def synthesize_speech(
    text: str, 
    voice_id: str, 
    model_id: str, 
    api_key: str,
    chunk_index: int,
    total_chunks: int
) -> bytes:
    """
    Synthesize speech from text using Inworld TTS API.
    
    Args:
        text: Text to synthesize
        voice_id: Voice ID to use
        model_id: Model ID to use
        api_key: API key for authentication
        chunk_index: Index of this chunk (for logging)
        total_chunks: Total number of chunks (for logging)
        
    Returns:
        bytes: Audio data
    """
    url = "https://api.inworld.ai/tts/v1/voice"
    
    headers = {
        "Content-Type": "application/json",
        "Authorization": f"Basic {api_key}"
    }
    
    request_data = {
        "text": text,
        "voice_id": voice_id,
        "model_id": model_id,
        "audio_config": {
            "audio_encoding": "LINEAR16",
            "sample_rate_hertz": SAMPLE_RATE
        }
    }
    
    try:
        print(f"[{chunk_index + 1}/{total_chunks}] Synthesizing chunk ({len(text)} chars)...")
        print(f"   Preview: \"{text[:60]}...\"")
        
        response = requests.post(url, headers=headers, json=request_data)
        response.raise_for_status()
        
        result = response.json()
        audio_data = base64.b64decode(result["audioContent"])
        
        print(f"   Done: {len(audio_data)} bytes")
        return audio_data
        
    except requests.exceptions.RequestException as e:
        print(f"HTTP Error for chunk {chunk_index + 1}: {e}")
        if hasattr(e, 'response') and e.response is not None:
            try:
                error_detail = e.response.json()
                print(f"   Error details: {error_detail}")
            except:
                print(f"   Response text: {e.response.text}")
        raise


def extract_raw_audio(audio_data: bytes) -> bytes:
    """Extract raw audio data from buffer (skip WAV header if present)."""
    if len(audio_data) > 44 and audio_data[:4] == b'RIFF':
        return audio_data[44:]
    return audio_data


def calculate_duration(raw_audio: bytes) -> float:
    """Calculate audio duration from raw PCM data."""
    bytes_per_second = SAMPLE_RATE * (BITS_PER_SAMPLE // 8) * CHANNELS
    return len(raw_audio) / bytes_per_second


def format_time(seconds: float) -> str:
    """Format seconds as MM:SS.mmm"""
    mins = int(seconds // 60)
    secs = seconds % 60
    return f"{mins}:{secs:06.3f}"


def combine_audio_buffers(
    audio_buffers: list[bytes], 
    chunks: list[TextChunk]
) -> tuple[bytes, list[SplicePoint], float]:
    """
    Combine multiple audio buffers and create splice report.
    
    Args:
        audio_buffers: List of audio data buffers
        chunks: Original text chunks with positions
        
    Returns:
        Tuple of (combined_audio, splice_points, total_duration)
    """
    splice_points = []
    current_time = 0.0
    raw_buffers = []
    
    for index, buffer in enumerate(audio_buffers):
        raw_audio = extract_raw_audio(buffer)
        duration = calculate_duration(raw_audio)
        
        if index > 0:
            splice_points.append(SplicePoint(
                splice_index=index,
                timestamp=current_time,
                formatted_time=format_time(current_time),
                chunk_start_char=chunks[index].start_char,
                chunk_end_char=chunks[index].end_char,
                text_preview=chunks[index].text[:50] + "..."
            ))
        
        current_time += duration
        raw_buffers.append(raw_audio)
    
    combined_audio = b''.join(raw_buffers)
    return combined_audio, splice_points, current_time


def save_audio_to_file(audio_data: bytes, output_file: str):
    """Save combined audio to WAV file."""
    with wave.open(output_file, "wb") as wf:
        wf.setnchannels(CHANNELS)
        wf.setsampwidth(BITS_PER_SAMPLE // 8)
        wf.setframerate(SAMPLE_RATE)
        wf.writeframes(audio_data)
    
    print(f"Audio saved to: {output_file}")


def print_splice_report(splice_points: list[SplicePoint], total_duration: float):
    """Print splice report for quality checking."""
    print("\nSPLICE REPORT - Check these timestamps for voice quality:")
    print("=" * 70)
    
    if not splice_points:
        print("   No splices - text was short enough for single request")
        return
    
    print(f"   Total splices: {len(splice_points)}")
    print(f"   Total duration: {format_time(total_duration)}\n")
    
    for idx, point in enumerate(splice_points):
        print(f"   Splice #{idx + 1}:")
        print(f"      Timestamp: {point.formatted_time}")
        print(f"      Character position: {point.chunk_start_char}")
        print(f"      Text: \"{point.text_preview}\"")
        print()
    
    print("   Tip: Listen to timestamps above to verify consistent voice quality")
    print("=" * 70)


def synthesize_chunk_wrapper(args):
    """Wrapper function for parallel synthesis."""
    chunk, voice_id, model_id, api_key, index, total = args
    return index, synthesize_speech(chunk.text, voice_id, model_id, api_key, index, total)


def main():
    """Main function to demonstrate long text TTS synthesis."""
    print("Inworld TTS Long Text Synthesis Example")
    print("=" * 50)
    
    # Check API key
    api_key = check_api_key()
    if not api_key:
        return 1
    
    # Configuration
    voice_id = "Edward"
    model_id = "inworld-tts-1-max"
    output_file = "synthesis_long_output.wav"
    
    # Read input text file
    script_dir = Path(__file__).parent
    input_file = script_dir / ".." / "tests-data" / "text" / "chapter1.txt"
    
    try:
        text = input_file.read_text(encoding="utf-8")
        print(f"Loaded text file: {input_file}")
        print(f"   Total characters: {len(text)}")
    except Exception as e:
        print(f"Error reading input file: {e}")
        return 1
    
    # Chunk the text
    print(f"\nChunking text (min {MIN_CHUNK_SIZE} chars, max {MAX_CHUNK_SIZE} chars per chunk)...")
    chunks = chunk_text(text)
    print(f"   Created {len(chunks)} chunks\n")
    
    # Display chunk info
    for i, chunk in enumerate(chunks):
        print(f"   Chunk {i + 1}: {len(chunk.text)} chars (positions {chunk.start_char}-{chunk.end_char})")
    print()
    
    try:
        start_time = time.time()
        
        # Synthesize all chunks in parallel
        print("Starting parallel TTS synthesis for all chunks...\n")
        
        # Prepare arguments for parallel execution
        synthesis_args = [
            (chunk, voice_id, model_id, api_key, i, len(chunks))
            for i, chunk in enumerate(chunks)
        ]
        
        # Use ThreadPoolExecutor for parallel requests
        audio_buffers = [None] * len(chunks)
        with ThreadPoolExecutor(max_workers=len(chunks)) as executor:
            futures = {
                executor.submit(synthesize_chunk_wrapper, args): args[4]  # args[4] is index
                for args in synthesis_args
            }
            
            for future in as_completed(futures):
                index, audio_data = future.result()
                audio_buffers[index] = audio_data
        
        # Combine audio (buffers are in correct order)
        print("\nCombining audio chunks...")
        combined_audio, splice_points, total_duration = combine_audio_buffers(audio_buffers, chunks)
        
        # Save to file
        save_audio_to_file(combined_audio, output_file)
        
        synthesis_time = time.time() - start_time
        
        # Print splice report
        print_splice_report(splice_points, total_duration)
        
        print(f"\nTotal synthesis time: {synthesis_time:.2f} seconds")
        print(f"Synthesis completed! Output file: {output_file}")
        print(f"   Audio duration: {format_time(total_duration)}")
        
    except Exception as e:
        print(f"\nSynthesis failed: {e}")
        return 1
    
    return 0


if __name__ == "__main__":
    exit(main())

File: js/example_tts_long_input.js
#!/usr/bin/env node
/**
 * Example script for Inworld TTS synthesis with long text input.
 *
 * This script demonstrates how to synthesize speech from long text by:
 * 1. Chunking text at natural boundaries (paragraphs → newlines → sentences)
 * 2. Processing chunks through the TTS API with controlled concurrency
 * 3. Stitching all audio outputs together
 * 4. Reporting splice points with timestamps for quality checking
 */

const fs = require('fs');
const axios = require('axios');
const path = require('path');

// Configuration
const INPUT_FILE_PATH = '../tests-data/text/chapter1.txt';  // Path to input text file (relative to this script)
const MIN_CHUNK_SIZE = 500;   // Minimum characters before looking for break point
const MAX_CHUNK_SIZE = 1900;  // Maximum chunk size (API limit is 2000)
const MAX_CONCURRENT_REQUESTS = 2;  // Limit parallel requests to avoid RPS limits
const MAX_RETRIES = 3;        // Maximum retries for rate limit errors
const RETRY_BASE_DELAY = 1000; // Base delay for exponential backoff (ms)

// Audio configuration
const SAMPLE_RATE = 48000;
const BITS_PER_SAMPLE = 16;
const CHANNELS = 1;
const SPLICE_BREAK = 0.5;     // Seconds of silence to insert between chunks for natural pause

/**
 * Check if INWORLD_API_KEY environment variable is set.
 * @returns {string|null} API key or null if not set
 */
function checkApiKey() {
    const apiKey = process.env.INWORLD_API_KEY;
    if (!apiKey) {
        console.log('Error: INWORLD_API_KEY environment variable is not set.');
        console.log('Please set it with: export INWORLD_API_KEY=your_api_key_here');
        return null;
    }
    return apiKey;
}

/**
 * Find the best break point in text, prioritizing:
 * 1. Paragraph breaks (\n\n)
 * 2. Line breaks (\n)
 * 3. Sentence endings (.!?)
 * 4. Last space (fallback)
 * 
 * @param {string} text - Text to search for break point
 * @param {number} minPos - Minimum position to start looking
 * @param {number} maxPos - Maximum position to look
 * @param {number} chunkIndex - Current chunk index for logging
 * @returns {number} Position of best break point
 */
function findBreakPoint(text, minPos, maxPos, chunkIndex) {
    const searchText = text.slice(0, maxPos);
    
    // 1. Try paragraph breaks (\n\n) after minPos
    let breakIndex = -1;
    let searchStart = minPos;
    
    while (true) {
        const idx = searchText.indexOf('\n\n', searchStart);
        if (idx === -1 || idx >= maxPos) break;
        if (idx >= minPos) {
            breakIndex = idx + 2; // Include the paragraph break
            break; // Use first paragraph break after minPos
        }
        searchStart = idx + 1;
    }
    
    if (breakIndex > 0) {
        console.log(`  Chunk ${chunkIndex + 1}: Found paragraph break at position ${breakIndex}`);
        return breakIndex;
    }
    
    // 2. Try single line breaks (\n) after minPos
    searchStart = minPos;
    while (true) {
        const idx = searchText.indexOf('\n', searchStart);
        if (idx === -1 || idx >= maxPos) break;
        if (idx >= minPos) {
            breakIndex = idx + 1;
            break;
        }
        searchStart = idx + 1;
    }
    
    if (breakIndex > 0) {
        console.log(`  Chunk ${chunkIndex + 1}: Found line break at position ${breakIndex}`);
        return breakIndex;
    }
    
    // 3. Try sentence endings after minPos
    const sentenceEndRegex = /[.!?]["'"']?\s+|[.!?]["'"']?$/g;
    let match;
    while ((match = sentenceEndRegex.exec(searchText)) !== null) {
        if (match.index >= minPos) {
            console.log(`  Chunk ${chunkIndex + 1}: Found sentence break at position ${match.index + match[0].length}`);
            return match.index + match[0].length;
        }
    }
    
    // 4. Fall back to any sentence end before maxPos
    sentenceEndRegex.lastIndex = 0;
    let lastSentenceEnd = -1;
    while ((match = sentenceEndRegex.exec(searchText)) !== null) {
        lastSentenceEnd = match.index + match[0].length;
    }
    if (lastSentenceEnd > 0) {
        console.log(`  Chunk ${chunkIndex + 1}: Found sentence break (fallback) at position ${lastSentenceEnd}`);
        return lastSentenceEnd;
    }
    
    // 5. Last resort: break at last space
    const spaceIndex = searchText.lastIndexOf(' ');
    const breakPos = spaceIndex > 0 ? spaceIndex + 1 : maxPos;
    console.log(`  Chunk ${chunkIndex + 1}: Found space break (fallback) at position ${breakPos}`);
    return breakPos;
}

/**
 * Chunk text into segments at natural boundaries.
 * Prioritizes paragraph breaks, then line breaks, then sentence endings.
 * 
 * @param {string} text - The full text to chunk
 * @returns {Array<{text: string, startChar: number, endChar: number}>} Array of text chunks
 */
function chunkText(text) {
    const chunks = [];
    let currentPosition = 0;
    
    while (currentPosition < text.length) {
        const remainingText = text.slice(currentPosition);
        
        // If remaining text is short enough, take it all
        if (remainingText.length <= MAX_CHUNK_SIZE) {
            const chunkContent = remainingText.trim();
            if (chunkContent.length > 0) {
                chunks.push({
                    text: chunkContent,
                    startChar: currentPosition,
                    endChar: text.length
                });
            }
            break;
        }
        
        // Find the best break point
        const chunkEnd = findBreakPoint(remainingText, MIN_CHUNK_SIZE, MAX_CHUNK_SIZE, chunks.length);
        
        const chunkContent = remainingText.slice(0, chunkEnd).trim();
        if (chunkContent.length > 0) {
            chunks.push({
                text: chunkContent,
                startChar: currentPosition,
                endChar: currentPosition + chunkEnd
            });
        }
        
        currentPosition += chunkEnd;
    }
    
    return chunks;
}

/**
 * Sleep for a specified duration.
 * @param {number} ms - Milliseconds to sleep
 * @returns {Promise<void>}
 */
function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}

/**
 * Synthesize speech from text using Inworld TTS API with retry logic.
 * 
 * @param {string} text - Text to synthesize
 * @param {string} voiceId - Voice ID to use
 * @param {string} modelId - Model ID to use
 * @param {string} apiKey - API key for authentication
 * @param {number} chunkIndex - Chunk index for logging
 * @param {number} totalChunks - Total number of chunks
 * @returns {Promise<Buffer>} Audio data
 */
async function synthesizeSpeech(text, voiceId, modelId, apiKey, chunkIndex, totalChunks) {
    const url = 'https://api.inworld.ai/tts/v1/voice';
    
    const headers = {
        'Content-Type': 'application/json',
        'Authorization': `Basic ${apiKey}`
    };
    
    const requestData = {
        text: text,
        voice_id: voiceId,
        model_id: modelId,
        audio_config: {
            audio_encoding: 'LINEAR16',
            sample_rate_hertz: SAMPLE_RATE
        }
    };
    
    for (let attempt = 0; attempt < MAX_RETRIES; attempt++) {
        try {
            console.log(`[${chunkIndex + 1}/${totalChunks}] Synthesizing chunk (${text.length} chars)...`);
            
            const response = await axios.post(url, requestData, { headers });
            const audioData = Buffer.from(response.data.audioContent, 'base64');
            
            console.log(`[${chunkIndex + 1}/${totalChunks}] Done - ${audioData.length} bytes`);
            return audioData;
            
        } catch (error) {
            const isRateLimit = error.response?.status === 429;
            const isLastAttempt = attempt === MAX_RETRIES - 1;
            
            if (isRateLimit && !isLastAttempt) {
                const delay = RETRY_BASE_DELAY * Math.pow(2, attempt);
                console.log(`[${chunkIndex + 1}/${totalChunks}] Rate limited, retrying in ${delay}ms...`);
                await sleep(delay);
                continue;
            }
            
            console.log(`Error for chunk ${chunkIndex + 1}: ${error.message}`);
            if (error.response) {
                try {
                    console.log(`   Error details: ${JSON.stringify(error.response.data)}`);
                } catch {
                    console.log(`   Response text: ${error.response.data}`);
                }
            }
            throw error;
        }
    }
}

/**
 * Process chunks with controlled concurrency.
 * @param {Array} chunks - Text chunks to process
 * @param {string} voiceId - Voice ID
 * @param {string} modelId - Model ID
 * @param {string} apiKey - API key
 * @returns {Promise<Array<Buffer>>} Audio buffers in order
 */
async function synthesizeAllChunks(chunks, voiceId, modelId, apiKey) {
    const results = new Array(chunks.length);
    const queue = chunks.map((chunk, index) => ({ chunk, index }));
    
    async function processNext() {
        while (queue.length > 0) {
            const { chunk, index } = queue.shift();
            results[index] = await synthesizeSpeech(
                chunk.text,
                voiceId,
                modelId,
                apiKey,
                index,
                chunks.length
            );
        }
    }
    
    // Start concurrent workers
    const workers = [];
    for (let i = 0; i < Math.min(MAX_CONCURRENT_REQUESTS, chunks.length); i++) {
        workers.push(processNext());
    }
    
    await Promise.all(workers);
    return results;
}

/**
 * Extract raw audio data from buffer (skip WAV header if present).
 * @param {Buffer} audioData - Audio data (may include WAV header)
 * @returns {Buffer} Raw PCM audio data
 */
function extractRawAudio(audioData) {
    if (audioData.length > 44 && audioData.subarray(0, 4).equals(Buffer.from('RIFF'))) {
        return audioData.subarray(44);
    }
    return audioData;
}

/**
 * Calculate audio duration from raw PCM data.
 * @param {Buffer} rawAudio - Raw PCM audio data
 * @returns {number} Duration in seconds
 */
function calculateDuration(rawAudio) {
    const bytesPerSecond = SAMPLE_RATE * (BITS_PER_SAMPLE / 8) * CHANNELS;
    return rawAudio.length / bytesPerSecond;
}

/**
 * Format seconds as MM:SS.mmm
 * @param {number} seconds - Duration in seconds
 * @returns {string} Formatted time string
 */
function formatTime(seconds) {
    const mins = Math.floor(seconds / 60);
    const secs = seconds % 60;
    return `${mins}:${secs.toFixed(3).padStart(6, '0')}`;
}

/**
 * Create silence buffer for splice breaks.
 * @param {number} durationSeconds - Duration of silence in seconds
 * @returns {Buffer} Silent audio buffer
 */
function createSilenceBuffer(durationSeconds) {
    const bytesPerSecond = SAMPLE_RATE * (BITS_PER_SAMPLE / 8) * CHANNELS;
    const numBytes = Math.floor(bytesPerSecond * durationSeconds);
    // Ensure even number of bytes for 16-bit audio
    const alignedBytes = numBytes - (numBytes % 2);
    return Buffer.alloc(alignedBytes, 0);
}

/**
 * Combine multiple audio buffers and create splice report.
 * Inserts SPLICE_BREAK seconds of silence between chunks for natural pauses.
 * @param {Array<Buffer>} audioBuffers - Array of audio buffers
 * @param {Array<Object>} chunks - Original text chunks with positions
 * @returns {{combinedAudio: Buffer, splicePoints: Array, totalDuration: number}}
 */
function combineAudioBuffers(audioBuffers, chunks) {
    const splicePoints = [];
    const combinedBuffers = [];
    let currentTime = 0;
    
    // Create silence buffer once if we need it
    const silenceBuffer = SPLICE_BREAK > 0 ? createSilenceBuffer(SPLICE_BREAK) : null;
    
    audioBuffers.forEach((buffer, index) => {
        const rawAudio = extractRawAudio(buffer);
        const duration = calculateDuration(rawAudio);
        
        // Add silence before this chunk (except for the first chunk)
        if (index > 0 && silenceBuffer) {
            combinedBuffers.push(silenceBuffer);
            currentTime += SPLICE_BREAK;
        }
        
        if (index > 0) {
            splicePoints.push({
                spliceIndex: index,
                timestamp: currentTime,
                formattedTime: formatTime(currentTime),
                chunkStartChar: chunks[index].startChar,
                chunkEndChar: chunks[index].endChar,
                textPreview: chunks[index].text.substring(0, 50) + '...'
            });
        }
        
        combinedBuffers.push(rawAudio);
        currentTime += duration;
    });
    
    return {
        combinedAudio: Buffer.concat(combinedBuffers),
        splicePoints,
        totalDuration: currentTime
    };
}

/**
 * Create WAV file header.
 * @param {number} dataSize - Size of audio data
 * @returns {Buffer} WAV header
 */
function createWavHeader(dataSize) {
    const header = Buffer.alloc(44);
    const bytesPerSample = BITS_PER_SAMPLE / 8;
    const blockAlign = CHANNELS * bytesPerSample;
    const byteRate = SAMPLE_RATE * blockAlign;
    
    header.write('RIFF', 0);
    header.writeUInt32LE(36 + dataSize, 4);
    header.write('WAVE', 8);
    header.write('fmt ', 12);
    header.writeUInt32LE(16, 16);
    header.writeUInt16LE(1, 20);
    header.writeUInt16LE(CHANNELS, 22);
    header.writeUInt32LE(SAMPLE_RATE, 24);
    header.writeUInt32LE(byteRate, 28);
    header.writeUInt16LE(blockAlign, 32);
    header.writeUInt16LE(BITS_PER_SAMPLE, 34);
    header.write('data', 36);
    header.writeUInt32LE(dataSize, 40);
    
    return header;
}

/**
 * Save audio to file.
 * @param {Buffer} audioData - Audio data
 * @param {string} outputFile - Output file path
 */
function saveAudioToFile(audioData, outputFile) {
    const wavHeader = createWavHeader(audioData.length);
    fs.writeFileSync(outputFile, Buffer.concat([wavHeader, audioData]));
    console.log(`Audio saved to: ${outputFile}`);
}

/**
 * Print splice report for quality checking.
 * @param {Array<Object>} splicePoints - Splice point info
 * @param {number} totalDuration - Total audio duration
 */
function printSpliceReport(splicePoints, totalDuration) {
    if (splicePoints.length === 0) {
        console.log('No splices - text was short enough for single request');
        return;
    }
    
    console.log(`\nSplice Report (${splicePoints.length} splices, duration: ${formatTime(totalDuration)}):`);
    splicePoints.forEach((point, idx) => {
        console.log(`  #${idx + 1} at ${point.formattedTime} - "${point.textPreview}"`);
    });
}

/**
 * Read input text from file.
 * @param {string} inputFile - Path to input file
 * @returns {string|null} Text content or null on error
 */
function readInputText(inputFile) {
    try {
        const text = fs.readFileSync(inputFile, 'utf-8');
        console.log(`Loaded: ${inputFile} (${text.length} chars)`);
        return text;
    } catch (error) {
        console.log(`Error reading input file: ${error.message}`);
        return null;
    }
}

/**
 * Main function - high-level orchestration only.
 */
async function main() {
    console.log('Inworld TTS Long Text Synthesis\n');
    
    // Setup
    const apiKey = checkApiKey();
    if (!apiKey) return 1;
    
    // Configuration - modify these for your use case
    const voiceId = 'Edward';
    const modelId = 'inworld-tts-1-max';
    const outputFile = 'synthesis_long_output.wav';
    const inputFile = path.join(__dirname, INPUT_FILE_PATH);
    
    // Read input text
    const text = readInputText(inputFile);
    if (!text) return 1;
    
    // Split into chunks
    const chunks = chunkText(text);
    console.log(`Split into ${chunks.length} chunks (min: ${MIN_CHUNK_SIZE}, max: ${MAX_CHUNK_SIZE} chars)\n`);
    
    try {
        const startTime = Date.now();
        
        // Synthesize all chunks
        console.log(`Synthesizing with ${MAX_CONCURRENT_REQUESTS} concurrent requests...\n`);
        const audioBuffers = await synthesizeAllChunks(chunks, voiceId, modelId, apiKey);
        
        // Combine audio
        console.log('\nCombining audio...');
        const { combinedAudio, splicePoints, totalDuration } = combineAudioBuffers(audioBuffers, chunks);
        
        // Save output
        saveAudioToFile(combinedAudio, outputFile);
        
        // Report
        printSpliceReport(splicePoints, totalDuration);
        
        const elapsed = (Date.now() - startTime) / 1000;
        console.log(`\nCompleted in ${elapsed.toFixed(2)}s - Audio duration: ${formatTime(totalDuration)}`);
        
    } catch (error) {
        console.log(`\nSynthesis failed: ${error.message}`);
        return 1;
    }
    
    return 0;
}

if (require.main === module) {
    main().then(process.exit);
}

module.exports = { chunkText, synthesizeSpeech, combineAudioBuffers, synthesizeAllChunks };
